IGEL_LIMIT_BAYTRAIL_C_STATES
	  Do not allow C states below 5 for baytrail CPUs. This should make them
	  more reliable and fix some nasty freezes.

	  Source: IGEL
diff --git a/drivers/idle/intel_idle.c b/drivers/idle/intel_idle.c
index 9ba83954c255..e6dc7e2cf26e 100644
--- a/drivers/idle/intel_idle.c
@@ -346,7 +349,13 @@ static struct cpuidle_state byt_cstates[] __initdata = {
 	{
 		.name = "C6N",
 		.desc = "MWAIT 0x58",
+#ifdef CONFIG_IGEL_LIMIT_BAYTRAIL_C_STATES
+		/* gottwald@igel.com Try to fix j1900 freezes
+		 * Patch taken from: https://bugzilla.kernel.org/show_bug.cgi?id=109051#c865 */
+		.flags = MWAIT2flg(0x58) | CPUIDLE_FLAG_TLB_FLUSHED | CPUIDLE_FLAG_UNUSABLE,
+#else
 		.flags = MWAIT2flg(0x58) | CPUIDLE_FLAG_TLB_FLUSHED,
+#endif
 		.exit_latency = 300,
 		.target_residency = 275,
 		.enter = intel_idle,
diff --git a/drivers/idle/intel_idle.c b/drivers/idle/intel_idle.c
index 9ba83954c255..e6dc7e2cf26e 100644
--- a/drivers/idle/intel_idle.c
@@ -354,7 +363,13 @@ static struct cpuidle_state byt_cstates[] __initdata = {
 	{
 		.name = "C6S",
 		.desc = "MWAIT 0x52",
+#ifdef CONFIG_IGEL_LIMIT_BAYTRAIL_C_STATES
+		/* gottwald@igel.com Try to fix j1900 freezes
+		 * Patch taken from: https://bugzilla.kernel.org/show_bug.cgi?id=109051#c865 */
+		.flags = MWAIT2flg(0x52) | CPUIDLE_FLAG_TLB_FLUSHED | CPUIDLE_FLAG_UNUSABLE,
+#else
 		.flags = MWAIT2flg(0x52) | CPUIDLE_FLAG_TLB_FLUSHED,
+#endif
 		.exit_latency = 500,
 		.target_residency = 560,
 		.enter = intel_idle,
